### PR TITLE
feat(core,api): IAM auth mode for HTTP requests (REST and GQL)

### PR DESCRIPTION
**Pull request by ragingsquirrel3**
This PR does the following:
* migrates logic for API key auth mode to auth provider, including definition in core
* consumes IAM auth provider in API plugin for HTTP requests, the result being that REST and GraphQL queries/mutations will work with IAM auth mode. I plan to have integration tests for this in another PR.
* unit tests for that stuff

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/api/amplify_api/lib/src/api_plugin_impl.dart [hunk 1]
```diff
@@ -98,8 +113,12 @@ class AmplifyAPIDart extends AmplifyAPI {
       type: EndpointType.graphQL,
       apiName: apiName,
     );
-    return _clientPool[endpoint.name] ??= AmplifyAuthorizationRestClient(
-        endpointConfig: endpoint.config, baseClient: _baseHttpClient);
+    return _clientPool[endpoint.name] ??= AmplifyHttpClient(
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/lib/src/api_plugin_impl.dart:116]: not really what this PR is about, but figured we want to use this client to have expected user agent content

**Review by dnys1 — APPROVED** [review]

File: packages/api/amplify_api/lib/src/decorators/authorize_http_request.dart [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+import 'package:http/http.dart' as http;
+import 'package:meta/meta.dart';
+
+/// Transforms an HTTP request according to auth providers that match the endpoint
+/// configuration.
+@internal
+Future<http.BaseRequest> authorizeHttpRequest(http.BaseRequest request,
+    {required AWSApiConfig endpointConfig,
+    required AmplifyAuthProviderRepository authProviderRepo}) async {
+  if (!request.headers.containsKey(AWSHeaders.authorization)) {
```
Comment (Equartey) [inline packages/api/amplify_api/lib/src/decorators/authorize_http_request.dart:27]: To help with readability, could this be a guard clause? Something like...

```
if (request.headers.containsKey(AWSHeaders.authorization)) {
    return request;
}

final authType = endpointConfig.authorizationType;
...
```

Comment (ragingsquirrel3) [reply packages/api/amplify_api/lib/src/decorators/authorize_http_request.dart:27]: yep, good suggestion! will do in next rev

File: packages/api/amplify_api/lib/src/api_plugin_impl.dart [hunk 3]
```diff
@@ -111,10 +130,12 @@ class AmplifyAPIDart extends AmplifyAPI {
       type: EndpointType.rest,
       apiName: apiName,
     );
-    return _clientPool[endpoint.name] ??= AmplifyAuthorizationRestClient(
+    return _clientPool[endpoint.name] ??= AmplifyHttpClient(
```
[the remark this answers is not in the document]
Comment (ragingsquirrel3) [reply packages/api/amplify_api/lib/src/api_plugin_impl.dart:135]: good call, done for next rev

File: packages/amplify_core/lib/src/types/common/amplify_auth_provider.dart [hunk 4]
```diff
@@ -34,6 +34,12 @@ class IamAuthProviderOptions extends AuthProviderOptions {
   const IamAuthProviderOptions({required this.region, required this.service});
 }
 
+class ApiKeyAuthProviderOptions extends AuthProviderOptions {
+  final String apiKey;
+
+  const ApiKeyAuthProviderOptions({required this.apiKey});
```
[the remark this answers is not in the document]
Comment (ragingsquirrel3) [reply packages/amplify_core/lib/src/types/common/amplify_auth_provider.dart:40]: done thanks!

**Review by Equartey — APPROVED** [review]
Looks good, nice work!

**Pull request merged**